### PR TITLE
seL4SharedDataWithCaps: Fix hardware_cached

### PR DESCRIPTION
**Pull request by hlyytine**
Signed-off-by: Hannu Lyytinen <hannux@ssrc.tii.ae>

### CODE REVIEW

**Comment by hlyytine** [comment]
Found out that hardware_cached attribute does not have any effect on ARM, the frames are always created with uncached attribute. This is a fix for that.

Speaking of which, I managed to make the guest Linux cross connector kernel module working properly by just changing the mapping type from UIO_MEM_PHYS (uncached) to UIO_MEM_IOVA (cached). Should we make cached the default setting, that way ARM and x86 would be unified behaviour? If yes, I can make the needed PRs here and at camkes-vm-linux.

**Comment by axel-h** [comment]
Seems a sane assumption to be that shared data is always cached, it's not meant to be used for peripheral access.

**Comment by hlyytine** [comment]
Augmented this PR with making the dataport mapping cached on ARM as well, see also https://github.com/seL4/camkes-vm-linux/pull/10

File: templates/seL4SharedDataWithCaps.template.c [hunk 1]
```diff
@@ -52,12 +52,12 @@
...
 
 /*- set arch_name = obj_space.spec.arch.capdl_name() -*/
-/*- if arch_name in ['ia32', 'x86_64'] -*/
+/*- if arch_name in ['aarch32', 'aarch64', 'arm_hyp', 'ia32', 'x86_64'] -*/
     /*- set cached_default = true -*/
 /*- else -*/
     /*- set cached_default = false -*/
 /*- endif -*/
-/*- set cached = configuration[me.instance.name].get('%s_hardware_cached' % me.instance.name, cached_default) -*/
+/*- set cached = configuration[me.instance.name].get('%s_hardware_cached' % me.interface.name, cached_default) -*/
```
Comment (axel-h) [inline templates/seL4SharedDataWithCaps.template.c:60]: Should this be a separate commit, as it seems to fix a bug?

Comment (hlyytine) [reply templates/seL4SharedDataWithCaps.template.c:60]: Did you mean separate PR, it is a separate commit already? Waiting for comments also from @kent-mcleod. Actually the other one, making the mapping cached, is kind of bug fix also, because I am quite confident vm-introspect example from camkes-vm-examples would not have been running OK on ARM. Might have worked for most of time, but has potential to misbehave, as is common when you mix cached and non-cached mappings.

Comment (axel-h) [reply templates/seL4SharedDataWithCaps.template.c:60]: It's just the change from `instance` to `interface` here that I was wondering about. Seem unrelated to any caching.

Comment (axel-h) [reply templates/seL4SharedDataWithCaps.template.c:60]: Ah no. just ignore my comment. Seem I was looking at the wrong commit.

**Review by wom-bat — APPROVED** [review]
This looks good,

**Comment by kent-mcleod** [comment]
@hlyytine Sorry, this seemed to fall off my radar. I believe this was originally setting Arm mappings to uncached because of the uncached mapping in the UIO module. My only worry is that changing the default caching behavior of the connector for Arm platforms which could lead to consistency errors for any existing project expecting the mappings to be uncached. However, I think we can just accept this as an acceptable breaking change considering that the connector is for shared memory between applications and not devices.

Thanks for fixing it!

File: templates/seL4SharedDataWithCaps.template.c [hunk 2]
```diff
@@ -52,12 +52,12 @@
 /*? macros.shared_buffer_symbol(sym=dataport_symbol_name, shmem_size=shmem_symbol_size, page_size=page_size) ?*/
 
 /*- set arch_name = obj_space.spec.arch.capdl_name() -*/
-/*- if arch_name in ['ia32', 'x86_64'] -*/
+/*- if arch_name in ['aarch32', 'aarch64', 'arm_hyp', 'ia32', 'x86_64'] -*/
```
Comment (axel-h) [inline templates/seL4SharedDataWithCaps.template.c:55]: Are you sure this works? I find `capdl_name()` implemented in `capdl/python-capdl-tool/capdl/util.py` and it returns `arm11` for `ARM32Arch`, independent of `hyp` being `true`or `false`. Thus, I wonder if this should check `options.architecture` instead of `obj_space.spec.arch.capdl_name()`?
